test/junit/scala/collection/mutable/MutationTrackingTest.scala

```diff
@@ -0,0 +1,142 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.collection
+package mutable
+
+import java.util.ConcurrentModificationException
+
+import org.junit.Test
+
+import scala.tools.testkit.AssertUtil.assertThrows
+
+class MutationTrackingTest {
+  import MutationTrackingTest._
+
+  @Test
+  def listBuffer(): Unit = {
+    val c = checker(ListBuffer).runSuites(Suite.buffer)
+    c checkThrows { _.mapInPlace(_ + 1) }
+    c checkThrows { _.flatMapInPlace(i => (i + 1) :: Nil) }
+    c checkThrows { _.filterInPlace(_ => true) }
+  }
+}
+
+private object MutationTrackingTest {
+  private def checker[C <: Iterable[Int]](factory: Factory[Int, C]): Checker[C] =
+    new Checker(factory).runSuites(Suite.nop)
+
+  /** Checks for exceptions when mutations happen. */
+  private class Checker[C <: Iterable[Int]](factory: Factory[Int, C]) {
+    private def runOp(op: C => Any): Unit = {
+      val coll = (factory.newBuilder += 1 += 2 += 3 += 4).result()
+      val it = coll.iterator
+      op(coll)
+      it.foreach(_ => ())
+    }
+
+    /** Checks that no exception is thrown by an operation. */
+    def checkFine(op: C => Any): Unit = runOp(op)
+
+    /** Checks that an exception is thrown by an operation. */
+    def checkThrows(op: C => Any): Unit = assertThrows[ConcurrentModificationException](runOp(op))
+
+    /** Run suites with this checker. */
+    def runSuites(suites: Suite[C]*): this.type = { suites foreach { _ runWith this }; this }
+  }
+
+  /** A suite of checks. */
+  private trait Suite[C <: Iterable[Int]] {
+    def runWith(c: Checker[C]): Unit
+  }
+
+  private object Suite {
```

Comment:
oops, I overengineered it. but it will be highly reusable, so it's worth it

Reply:
I would like that on a t-shirt. I'm not sure which part goes on the front and which part goes on the back.